syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+    bigint2ts)
+from syncstorage.storage.sql.dbconnect import MAX_TTL, bso
+from syncstorage.storage.sql.queries_spanner import (
+    COLLECTION_CURRENT_TIMESTAMP,
+    COLLECTIONS_SIZES,
+    COLLECTIONS_COUNTS,
+    FIND_ITEMS,
+    STORAGE_SIZE,
+)
+
+logger = logging.getLogger(__name__)
+
+INT64_MAX = 63 ** 1 - 1
+
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+
+
+def dt2ts(dt):
+    """Convert a Python datetime to seconds"""
+    val = (dt.replace(tzinfo=None) - EPOCH).total_seconds()
+    return get_timestamp(math.floor(val * 100) / 100)
+
+
+def nsint2ts(nsint):
+    """Convert Spanner TIMESTAMP values of nanoseconds to seconds"""
+    return get_timestamp(nsint / 1000.0 / 1000.0 / 1000.0)
+
+
+def ts2dt(ts):
+    """Convert a timestamp to Python datetime"""
+    return datetime.datetime.utcfromtimestamp(ts)
+
+
+def with_session(func):
+    @functools.wraps(func)
+    def with_session_wrapper(self, *args, **kwds):
+        if hasattr(self._tldata, "session"):
+            return func(self, self._tldata.session, *args, **kwds)
+        with SpannerStorageSession(self) as spanner_session:
+            with SessionCheckout(self._pool) as session:
+                with session.transaction() as transaction:
+                    spanner_session.set_transaction(transaction)
+                    return func(self, self._tldata.session, *args, **kwds)
+
+    return with_session_wrapper
+
+
+def user_key(user):
+    """Create a compound user-key using the fxa_uid and fxa_kid
+
+    Joined here rather than used as a compound primary key to allow re-use
+    of existing queries.
+
+    """
+    return ":".join([user["fxa_uid"], user["fxa_kid"]])
+
+
+def getq(query):
+    kwargs = {}
+    if "%(bso)s" in query:
+        kwargs["bso"] = "bso"
+    if "%(bui)s" in query:
+        raise RuntimeError("XXX: batches")
+    if "%(ids)s" in query:
+        kwargs["ids"] = "@ids"
+    if kwargs:
+        query = query % kwargs
+    return query.replace(":", "@")
+
+
+class SpannerStorage(SyncStorage):
+    """Storage plugin implemented using a Spanner database.
+
+    This class implements the storage plugin API using Google Spanner.  You
+    must specify the Spanner database URI string to connect to, as follows:
+
+        spanner://INSTANCE_ID:DATABASE_ID
+
+    """
+
+    def __init__(self, sqluri, standard_collections=False, **dbkwds):
+        self.sqluri = sqluri
+        instance_id, database_id = sqluri[len("spanner://"):].split(":")
+        self.client = spanner.Client()
+        self._instance = self.client.instance(instance_id)
+        self._pool = pool = spanner.BurstyPool()
+        self._database = self._instance.database(database_id, pool=pool)
+
+        # There doesn't seem to be a reliable cross-database way to set the
+        # initial value of an autoincrement column.
+        self.standard_collections = standard_collections
+        if self.standard_collections and dbkwds.get("create_tables", False):
+            pass
```

Comment:
If this is still WIP, may be worth throwing here or something rather than just `pass`ing?

Reply:
Going to throw an error, as the database should be setup externally.